pkg/collection/linkedlist.go2

```diff
@@ -0,0 +1,189 @@
+package collection
+
+//node of linkedlist
+type LinkedListNode[E any] struct {
+	Val        E
+	Prev, Next *LinkedListNode[E]
+}
+
+//linkedlist is a circular doubly linked list
+//linkedlist has two dummy nodes, one is head dummyNode, the other is tail dummyNode
+type LinkedList[E any] struct {
+	//一个头节点一个尾节点
+	head, tail *LinkedListNode[E]
+	//length of linkedlist
+	len int
+}
+
+func InitLinkedList[E any]() *LinkedList[E] {
+	h, t := &LinkedListNode[E]{}, &LinkedListNode[E]{}
+	h.Prev, h.Next, t.Prev, t.Next = t, t, h, h
+	return &LinkedList[E]{
+		h,
+		t,
+		0,
+	}
+}
+
+func (l *LinkedList[E]) Len() int {
+	return l.len
+}
+
+func (l *LinkedList[E]) Empty() bool {
+	return l.len == 0
+}
+
+func (l *LinkedList[E]) Clear() {
+	l.len = 0
+	l.head.Next, l.head.Prev, l.tail.Prev, l.tail.Next = l.tail, l.tail, l.head, l.head
+}
+
+func (l *LinkedList[E]) Traverse() []E {
+	ret := make([]E, l.len)
+	cur := l.head.Next
+	for i := 0; i < l.Len(); i++ {
+		ret[i], cur = cur.Val, cur.Next
+	}
+	return ret
+}
+
+//add a linkedlist node to the given index
+//note: won't do anything if the index is illegal
+//If the index is a non-negative number, the index is [0, l.len],
+//which means that the value is added to the position where the index is index
+//If the index is a negative number, the index is [-1,-l.len-1],
+//-index means the I-th node from the tail
+//for example, -1 means the prev node of the dummy tail node.
+func (l *LinkedList[E]) Add(val E, index int) {
+	//check if index is illegal
+	if index > l.len || index < -l.len-1 {
+		return
+	}
+
+	if index >= 0 {
+		l.addFromFirst(val, index)
+	} else {
+		l.addFromLast(val, -index)
+	}
+
+	l.len++
+}
+
+func (l *LinkedList[E]) addFromFirst(val E, index int) {
+	cur := l.head
+	for i := 0; i < index; i++ {
+		cur = cur.Next
+	}
+	node := &LinkedListNode[E]{
+		Val: val,
+	}
+	node.Prev, node.Next, cur.Next, cur.Next.Prev = cur, cur.Next, node, node
```

Comment:
If index = 0,
cur = head,
After running this statement, `node.Prev = node`. 
It seems that we don't need `cur.Next.Prev = node`